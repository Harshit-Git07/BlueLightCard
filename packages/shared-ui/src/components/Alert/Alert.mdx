import { Meta, Canvas, Controls } from '@storybook/blocks';

import * as Alert from './Alert.stories';

<Meta of={Alert} />

# Alert

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)

## Overview

The Alert component displays important messages to users in different formats. It can be configured as a **full-width banner** that appears at the top of the page, ensuring visibility while pushing other content down, or as an **inline alert** within specific sections of the page. The component is customisable and can convey different states of information using distinct colors and icons.

### Prop Details

<Controls />

## Extra info

➡️ You can customise the alert's appearance with the `alertBackgroundColor` prop, utilising token colors to maintain design consistency across your application.

➡️ Use the `iconAccentColor` prop to define the color of the alert icon, again leveraging token colors for a cohesive visual style.

## Variants

The Alert component can be rendered in different styles depending on its variant.

### Banner Alert

The **Banner Alert** is a sticky, full-width alert that remains at the top of the page. It ensures that important messages are prominently displayed, capturing user attention without being overly intrusive. This variant is ideal for critical notifications that require immediate user action or awareness. It can be dismissed by users if the `isDismissable` prop is set.

<Canvas of={Alert.Banner} />

### Inline Alert

The **Inline Alert** is designed to provide contextual information within specific sections of a page. It does not disrupt the overall layout, making it suitable for conveying messages related to specific content or actions. This alert type is perfect for non-intrusive notifications that still need to be acknowledged by users.

For demonstration purposes, the Inline Alert examples are wrapped in a fixed width of 500 pixels. This ensures that the layout remains consistent and visually appealing, allowing you to see how the Inline Alert behaves within a confined space.

<Canvas of={Alert.Inline} />

## Call To Action

Both the **Banner Alert** and **Inline Alert support** customisable call-to-action (CTA) elements by allowing you to pass buttons, links, or other interactive components as children. This provides flexibility to tailor the alert's functionality and appearance to your specific needs.

## Button Alert

The Button Alert can include a button passed as a child to the Alert component, allowing flexibility for call-to-action (CTA). This button can be fully customised through the children prop, making it adaptable to various use cases. The button is a Primary button example.

## <Canvas of={Alert.BannerButton} /> <Canvas of={Alert.InlineButton} />

## Link Alert

The Link Alert includes a link passed as a child through the children prop. This gives developers
the ability to customise the content inside the alert while maintaining layout flexibility for links
or other interactive elements. The link is a Tertiary button example.

## <Canvas of={Alert.BannerLink} /> <Canvas of={Alert.InlineLink} />

## States

### Success Alert

The **Success Alert** indicates that an action has been successfully completed. It provides users with positive feedback, reinforcing their actions and encouraging continued engagement. This alert can also offer additional details about the successful operation.

<Canvas of={Alert.AlertSuccess} />

### Info Alert

The **Info Alert** is used to provide users with useful information that may help in their interactions. This alert type is non-intrusive and typically includes helpful tips or important notes that do not require immediate action but enhance user understanding.

<Canvas of={Alert.AlertInfo} />

### Warning Alert

The **Warning Alert** serves to inform users of potential issues that require attention. It acts as a cautionary notice, prompting users to verify their actions or entries to avoid possible mistakes. This alert can guide users towards making better decisions.

<Canvas of={Alert.AlertWarning} />

### Error Alert

The **Error Alert** communicates a problem that has occurred during a process or action. It clearly indicates that something went wrong and usually provides guidance on how to resolve the issue. This alert is critical for user awareness and helps them navigate errors effectively.

<Canvas of={Alert.AlertError} />

# Rich Text Alert

The `Rich Text Alert` component is a versatile alert that communicates important information effectively. It can be rendered as either a full-width banner or inline, making it suitable for various notifications and announcements. This design captures user attention in a visually appealing way.

The `subtext` of this example showcases rich text and includes:

- A bolded first line: `line one`, emphasised for importance.
- An italicised second line: `line two`, providing additional context.
- A clickable link, `"View details"`, which enables users to navigate for more information, utilising the `Link` or `Paragraph` components as needed.

  <Canvas of={Alert.RichTextBanner} /> <Canvas of={Alert.RichTextInline} />

## Custom Alert

The **Custom Alert** is a flexible alert component that provides customisable styles and behaviors. This example demonstrates a banner alert with a specific background color and icon accent color to enhance visibility and aesthetic appeal.

<Canvas of={Alert.CustomBanner} /> <Canvas of={Alert.CustomInline} />
