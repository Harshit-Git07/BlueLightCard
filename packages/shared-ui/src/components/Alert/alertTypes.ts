type BaseAlertProps = {
  title: string;
  subtext?: React.ReactNode;
  iconAccentColor?: string;
  alertBackgroundColor?: string;
  children?: React.ReactNode;
  isFullWidth?: boolean;
};

type DefaultStateAlertProps = BaseAlertProps & {
  state: 'Default';
  icon: string;
};

type SystemStateAlertProps = BaseAlertProps & {
  state: Exclude<State, 'Default'>;
  icon?: string;
};

type BannerAlertProps = (DefaultStateAlertProps | SystemStateAlertProps) & {
  variant: 'Banner';
  isDismissable?: boolean;
};

type InlineAlertProps = (DefaultStateAlertProps | SystemStateAlertProps) & {
  variant: 'Inline';
};

export type AlertProps = BannerAlertProps | InlineAlertProps;

export interface ColorConfig {
  iconColor: string;
  backgroundColor: string;
}

export interface AlertColorConfigProps {
  [key: string]: ColorConfig;
}

export type State = 'Default' | 'Success' | 'Info' | 'Warning' | 'Error';

export type Variant = 'Banner' | 'Inline';
