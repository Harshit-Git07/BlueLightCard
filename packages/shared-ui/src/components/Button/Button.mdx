import { Meta, Canvas, Controls } from '@storybook/blocks';

import * as Button from './Button.stories';

<Meta of={Button} />

# Button

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)

## Overview
Overview description of the button component...

## Props
<Controls />

## Usage
➡️ Usage instruction one

➡️ Usage instruction two

➡️ Usage instruction three

➡️ Usage instruction four

## Variants
List of variants can be shown below here.

### Default
<Canvas of={Button.Default} />

### Primary
<Canvas of={Button.Primary} />

### Secondary
<Canvas of={Button.Secondary} />

### Tertiary
<Canvas of={Button.Tertiary} />

### Link
<Canvas of={Button.Link} />

### Primary Inverted
<Canvas of={Button.PrimaryInverted} />

### Secondary Inverted
<Canvas of={Button.SecondaryInverted} />
