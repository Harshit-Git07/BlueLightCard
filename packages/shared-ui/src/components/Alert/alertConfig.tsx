import { AlertColorConfigProps, State } from './alertTypes';

import { fas, IconDefinition } from '@fortawesome/pro-solid-svg-icons';
import { IconProp } from '@fortawesome/fontawesome-svg-core';

export const alertColorConfig: AlertColorConfigProps = {
  Default: {
    iconColor: 'text-colour-onSurface-light dark:text-colour-onSurface-dark',
    backgroundColor: 'bg-colour-surface-container-light dark:bg-colour-surface-container-dark',
  },
  Success: {
    iconColor: 'text-colour-success-light dark:text-colour-success-dark',
    backgroundColor: 'bg-colour-success-bright-light dark:bg-colour-success-container-dark',
  },
  Info: {
    iconColor: 'text-colour-onInfo-container-light dark:text-colour-onInfo-container-dark',
    backgroundColor: 'bg-colour-info-bright-light dark:bg-colour-info-bright-dark',
  },
  Warning: {
    iconColor: 'text-colour-onWarning-container-light dark:text-colour-onWarning-container-dark',
    backgroundColor: 'bg-colour-warning-bright-light dark:bg-colour-warning-bright-dark',
  },
  Error: {
    iconColor: 'text-colour-onError-container-light dark:text-colour-onError-container-dark',
    backgroundColor: 'bg-colour-error-bright-light dark:bg-colour-error-bright-dark',
  },
};

const iconMapper: { [key in State]: IconDefinition | null } = {
  Default: null,
  Success: fas.faCheckCircle,
  Info: fas.faInfoCircle,
  Warning: fas.faExclamationTriangle,
  Error: fas.faTimesCircle,
};

export const getIconByStateOrString = (state: State, icon?: string): IconProp | null => {
  if (icon && fas[icon as keyof typeof fas]) {
    return fas[icon as keyof typeof fas];
  }

  if (state && iconMapper[state] !== null) {
    return iconMapper[state];
  }

  return null;
};
