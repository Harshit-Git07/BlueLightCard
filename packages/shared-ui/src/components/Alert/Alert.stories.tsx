import { Meta, StoryFn } from '@storybook/react';
import Alert from '.';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faCircle } from '@fortawesome/pro-solid-svg-icons';

import { ThemeVariant } from 'src/types';
import Button from '../Button';
import Link from '../Link';
const icons = { faCircle };

library.add(...Object.values(icons));
const componentMeta: Meta<typeof Alert> = {
  title: 'Component System/Alert',
  component: Alert,
  argTypes: {
    isDismissable: {
      control: 'boolean',

      if: { arg: 'variant', eq: 'Banner' },
    },
    icon: {
      control: {
        type: 'text',
      },
      description: 'Custom icon to display (required for Default state) *',
    },
    variant: {
      control: 'select',
      options: ['Banner', 'Inline'],
    },
  },
};

const AlertTemplate: StoryFn<typeof Alert> = (args) => <Alert {...args} />;

export const Banner = AlertTemplate.bind({});
Banner.args = {
  variant: 'Banner',
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a banner alert on desktop',
  isDismissable: true,
};

export const Inline = AlertTemplate.bind({});
Inline.args = {
  variant: 'Inline',
  icon: 'faCircle',
  state: 'Default',
  title: 'This is an inline alert on desktop',
  subtext: <p>Some subtext can go here it can be up to two lines.</p>,
};

Inline.argTypes = {
  isDismissable: { table: { disable: true } },
};

Inline.decorators = [
  (Story) => (
    <div style={{ width: '500px', margin: '0 auto' }}>
      <Story />
    </div>
  ),
];

export const BannerButton = AlertTemplate.bind({});

BannerButton.args = {
  variant: 'Banner',
  state: 'Default',
  icon: 'faCircle',
  title:
    'This is a Banner button alert with very long title-  it has no sub text and no close but has a cta and maintains minimum height of 84px.',
  isDismissable: true,
  children: (
    <Button
      type="button"
      iconLeft={icons.faCircle}
      iconRight={icons.faCircle}
      variant={ThemeVariant.Primary}
      onClick={() => {}}
    >
      Click me
    </Button>
  ),
};

export const InlineButton = AlertTemplate.bind({});

InlineButton.args = {
  variant: 'Inline',
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a Inline button alert',
  subtext: <p>Some subtext can go here it can be up to two lines.</p>,
  children: (
    <Button
      type="button"
      iconLeft={icons.faCircle}
      iconRight={icons.faCircle}
      variant={ThemeVariant.Primary}
      onClick={() => {}}
    >
      Click me
    </Button>
  ),
};

InlineButton.decorators = [
  (Story) => (
    <div style={{ width: '500px', margin: '0 auto' }}>
      <Story />
    </div>
  ),
];

export const BannerLink = AlertTemplate.bind({});

BannerLink.args = {
  variant: 'Banner',
  isDismissable: true,
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a Banner alert with link',
  children: (
    <Button type="button" variant={ThemeVariant.Tertiary} onClick={() => {}}>
      Click me
    </Button>
  ),
};

export const InlineLink = AlertTemplate.bind({});

InlineLink.args = {
  variant: 'Inline',
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a inline alert with link',
  children: (
    <Button type="button" variant={ThemeVariant.Tertiary} onClick={() => {}}>
      Click me
    </Button>
  ),
};

InlineLink.decorators = [
  (Story) => (
    <div style={{ width: '500px', margin: '0 auto' }}>
      <Story />
    </div>
  ),
];

export const AlertSuccess = AlertTemplate.bind({});

AlertSuccess.args = {
  variant: 'Banner',
  title: 'Success! Your action was completed.',
  state: 'Success',
  subtext: <p>Everything went smoothly. You can continue with your tasks.</p>,
  isDismissable: true,
};

export const AlertInfo = AlertTemplate.bind({});
AlertInfo.args = {
  variant: 'Banner',
  title: 'Information Alert',
  state: 'Info',
  subtext: <p>Please note that some important information is available.</p>,
  isDismissable: true,
};

export const AlertWarning = AlertTemplate.bind({});

AlertWarning.args = {
  variant: 'Banner',
  title: 'Warning! Check your input.',
  state: 'Warning',
  subtext: <p>There seems to be an issue. Please verify your entries.</p>,
  isDismissable: true,
};

export const AlertError = AlertTemplate.bind({});

AlertError.args = {
  variant: 'Banner',
  title: 'Error! Something went wrong.',
  state: 'Error',
  subtext: <p>An error occurred while processing your request. Please try again.</p>,
  isDismissable: true,
};

export const RichTextBanner = AlertTemplate.bind({});

RichTextBanner.args = {
  variant: 'Banner',
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a banner alert on desktop',
  subtext: (
    <>
      <p>
        This is {''}
        <strong className="text-colour-secondary-light dark:text-colour-secondary-dark">
          line one
        </strong>
      </p>
      <p>
        This is <i>line two</i>
      </p>
      <p className="pt-1">
        <Link onClickLink={() => {}} useLegacyRouting={false}>
          View details
        </Link>
      </p>
    </>
  ),
  isDismissable: true,
};

export const RichTextInline = AlertTemplate.bind({});

RichTextInline.args = {
  variant: 'Inline',
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a Inline alert on desktop',
  subtext: (
    <>
      <p>
        This is {''}
        <strong className="text-colour-secondary-light dark:text-colour-secondary-dark">
          line one
        </strong>
      </p>
      <p>
        This is <i>line two</i>
      </p>
      <p className="pt-1">
        <Link onClickLink={() => {}} useLegacyRouting={false}>
          View details
        </Link>
      </p>
    </>
  ),
};

RichTextInline.decorators = [
  (Story) => (
    <div style={{ width: '500px', margin: '0 auto' }}>
      <Story />
    </div>
  ),
];

export const CustomBanner = AlertTemplate.bind({});

CustomBanner.args = {
  variant: 'Banner',
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a banner alert on desktop',
  isDismissable: true,
  alertBackgroundColor:
    'bg-colour-secondary-highlight-fixed-light dark:bg-colour-secondary-highlight-fixed-dark',
  iconAccentColor:
    'text-colour-primary-highlight-fixed-light dark:text-colour-primary-highlight-fixed-dark',
};

export const CustomInline = AlertTemplate.bind({});

CustomInline.args = {
  variant: 'Inline',
  state: 'Default',
  icon: 'faCircle',
  title: 'This is a Inline alert on desktop',
  alertBackgroundColor:
    'bg-colour-secondary-highlight-fixed-light dark:bg-colour-secondary-highlight-fixed-dark',
  iconAccentColor:
    'text-colour-primary-highlight-fixed-light dark:text-colour-primary-highlight-fixed-dark',
};

CustomInline.decorators = [
  (Story) => (
    <div style={{ width: '500px', margin: '0 auto' }}>
      <Story />
    </div>
  ),
];

export default componentMeta;
