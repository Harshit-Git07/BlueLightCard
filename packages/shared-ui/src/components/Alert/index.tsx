import { FC, useMemo, useState } from 'react';
import { alertColorConfig, getIconByStateOrString } from './alertConfig';
import { faXmark } from '@fortawesome/pro-regular-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { conditionalStrings } from '../../utils/conditionalStrings';
import { AlertProps, State } from './alertTypes';
import { fonts } from '../../tailwind/theme';

const Alert: FC<AlertProps> = ({
  variant = 'Banner',
  state = 'Default',
  title,
  subtext,
  icon,
  iconAccentColor,
  alertBackgroundColor,
  children,
  isFullWidth,
  isDismissable = true,
}) => {
  const [visible, setVisible] = useState(true);

  const { iconColor, backgroundColor } = alertColorConfig[state];

  const resolvedIcon = getIconByStateOrString(state as State, icon);

  const backgroundAlertColor = alertBackgroundColor ?? backgroundColor;

  const iconAlertColor =
    iconAccentColor && iconAccentColor.length > 1 ? iconAccentColor : iconColor;

  const styleVariant = {
    Banner: {
      fullWidth: 'w-full',
      variant: 'sticky top-0 left-0 z-50 px-[1rem] tablet:px-[3.5rem] laptop:px-[6.5rem] py-[1rem]',
      layout: 'tablet:flex tablet:mx-6',
      title: `${fonts.titleSmall} tablet:font-typography-title-medium-semibold tablet:font-typography-title-medium-semibold-weight tablet:text-typography-title-medium-semibold tablet:leading-typography-title-medium-semibold tablet:tracking-typography-title-medium-semibold`,
      subtext: `pt-2 ${fonts.bodySmall} tablet:font-typography-body tablet:font-typography-body-weight tablet:text-typography-body tablet:leading-typography-body tablet:tracking-typography-body`,
      children: 'mt-[0.75rem] tablet:mt-0 tablet:justify-end',
      icon: 'tablet:items-center',
      iconSize: 'pt-1 tablet:pt-[0px] tablet:text-[24px]',
    },
    Inline: {
      fullWidth: isFullWidth ? 'w-full' : 'mx-auto',
      variant: 'relative  px-3 py-2',
      title: fonts.bodySmallSemiBold,
      subtext: `pt-1 ${fonts.bodySmall}`,
      layout: 'flex flex-col gap-0',
      children: 'flex flex-col gap-1',
      icon: '',
      iconSize: `pt-1 tablet:pt-[5px]`,
    },
  };

  const variantStyle = styleVariant[variant];

  const hideAlert = !visible ? 'hidden' : '';

  const childContainer = useMemo(() => {
    if (!children) return null;

    const flexDirectionStyles = variant === 'Inline' ? 'flex-col' : 'flex-row';
    const bannerStyles = variant === 'Banner' ? 'ml-auto mr-4 items-center' : '';
    const childContainerStyles = `${variantStyle.children} ${flexDirectionStyles} ${bannerStyles} flex justify-start items-start max-w-[250px] text-base w-full`;
    return <div className={childContainerStyles}>{children}</div>;
  }, [variantStyle.children, variant]);

  if (!visible) return null;

  const contentClass = conditionalStrings({
    'flex-col content-center  max-w-[750px]': true,
    'tablet:pr-[16px]': isDismissable,
  });

  return (
    <div
      className={`
        ${hideAlert} 
        ${variantStyle.fullWidth}
        ${backgroundAlertColor} 
        ${variantStyle.variant} 
        flex flex-col rounded justify-between 
      `}
      data-testid="alert"
    >
      <div className="flex basis-full flex-row" data-testid={`alert${variant}`}>
        {resolvedIcon && typeof resolvedIcon !== 'string' ? (
          <div className={`flex items-start ${variantStyle.icon}`}>
            <div className={`${iconAlertColor}`} data-testid="alert-icon">
              <FontAwesomeIcon
                data-testid="icon"
                icon={resolvedIcon}
                className={`text-[20px] flex content-start ${variantStyle.iconSize}`}
              />
            </div>
          </div>
        ) : null}
        <div className={`${variantStyle.layout} ml-4 m-auto w-full`}>
          <div className={contentClass}>
            <p
              className={`w-full text-colour-onSurface-light dark:text-colour-onSurface-dark break-words ${variantStyle.title}`}
            >
              {title}
            </p>

            {subtext && (
              <div
                className={`flex-col w-full ${variantStyle.subtext} text-colour-onSurface-subtle-light dark:text-colour-onSurface-subtle-dark break-words`}
              >
                {subtext}
              </div>
            )}
          </div>
          {childContainer}
        </div>
        {variant === 'Banner' && isDismissable && (
          <div className="flex items-start  tablet:items-center">
            <button
              onClick={() => setVisible(false)}
              aria-label="Close"
              className="content-center text-center rounded-full text-colour-onSurface-subtle-light dark:text-colour-onSurface-subtle-dark"
            >
              <FontAwesomeIcon icon={faXmark} className="text-[20px]" />
            </button>
          </div>
        )}
      </div>
    </div>
  );
};
export default Alert;
