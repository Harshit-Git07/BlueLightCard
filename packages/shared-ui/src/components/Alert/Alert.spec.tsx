import { render, screen, fireEvent } from '@testing-library/react';
import Alert from './index';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { fas, IconDefinition } from '@fortawesome/pro-solid-svg-icons';
import { State } from './alertTypes';
import Button from '../Button';

describe('Alert Component', () => {
  test('renders correctly for Banner variant', () => {
    const tree = renderer
      .create(
        <Alert
          variant="Banner"
          state="Success"
          title="This is a banner alert!"
          subtext={<p>Operation completed successfully.</p>}
          isDismissable={true}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly for Inline variant', () => {
    const tree = renderer
      .create(
        <Alert
          variant="Inline"
          state="Info"
          title="This is an inline alert!"
          subtext={<p>Some important information here.</p>}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('displays the correct title and subtext for Banner alert', () => {
    render(
      <Alert
        variant="Banner"
        state="Warning"
        title="Warning!"
        subtext={<p>Please check your input.</p>}
        isDismissable={true}
      />,
    );

    const bannerElement = screen.getByTestId('alertBanner');
    expect(bannerElement).toBeInTheDocument();

    expect(screen.getByText('Warning!')).toBeInTheDocument();
    expect(screen.getByText('Please check your input.')).toBeInTheDocument();
  });

  test('does not render dismiss button for Inline alert', () => {
    render(
      <Alert variant="Inline" state="Info" title="Information" subtext="This is an info alert." />,
    );

    const inlineElement = screen.getByTestId('alertInline');
    expect(inlineElement).toBeInTheDocument();

    expect(screen.queryByLabelText('Close')).not.toBeInTheDocument();
  });

  test('dismisses the Banner alert when the dismiss button is clicked', () => {
    render(
      <Alert
        variant="Banner"
        state="Error"
        title="Error!"
        subtext={<p>An error occurred.</p>}
        isDismissable={true}
      />,
    );

    const bannerElement = screen.getByTestId('alert');
    expect(bannerElement).toBeInTheDocument();

    expect(screen.getByText('Error!')).toBeInTheDocument();

    fireEvent.click(screen.getByLabelText('Close'));

    expect(screen.queryByText('Error!')).toBeNull();
  });

  const iconTests: Array<{ state: Exclude<State, 'Default'>; expectedIcon: IconDefinition }> = [
    { state: 'Success', expectedIcon: fas.faCheckCircle },
    { state: 'Info', expectedIcon: fas.faInfoCircle },
    { state: 'Warning', expectedIcon: fas.faExclamationTriangle },
    { state: 'Error', expectedIcon: fas.faTimesCircle },
  ];

  iconTests.forEach(({ state, expectedIcon }) => {
    test(`renders the correct icon for ${state} alert`, () => {
      render(
        <Alert
          variant="Banner"
          state={state}
          title={`${state} Alert`}
          subtext={<p>This is a ${state.toLowerCase()} alert.</p>}
        />,
      );

      const icon = screen.getByTestId('icon');
      expect(icon).toBeInTheDocument();
      expect(icon).toHaveAttribute('data-icon', expectedIcon.iconName);
    });
  });

  test('renders banner default state and mandatory custom icon and title, including strong and italic text', () => {
    render(
      <Alert
        variant="Banner"
        icon="faCircle"
        state="Default"
        title="Default Banner with icon and title"
        subtext={
          <>
            <p>
              This is the <strong>first line</strong>
            </p>
            <p>
              This is the <i>second line</i>
            </p>
          </>
        }
      />,
    );

    const bannerElement = screen.getByTestId('alertBanner');
    expect(bannerElement).toBeInTheDocument();

    const title = screen.getByText('Default Banner with icon and title');
    expect(title).toBeInTheDocument();

    const subtextOne = screen.getByText((content, element) => {
      return element?.textContent === 'This is the first line';
    });
    expect(subtextOne).toBeInTheDocument();

    const subtextTwo = screen.getByText((content, element) => {
      return element?.textContent === 'This is the second line';
    });
    expect(subtextTwo).toBeInTheDocument();

    const strongText = screen.getByText('first line');
    expect(strongText.tagName).toBe('STRONG');

    const italicText = screen.getByText('second line');
    expect(italicText.tagName).toBe('I');

    const icon = screen.getByTestId('icon');
    expect(icon).toHaveAttribute('data-icon', 'circle');
  });

  test('renders inline default state and mandatory custom icon and title, including strong and italic text', () => {
    render(
      <Alert
        variant="Inline"
        icon="faCircle"
        state="Default"
        title="Default Banner with icon and title"
        subtext={
          <>
            <p>
              This is the <strong>first line</strong>
            </p>
            <p>
              This is the <i>second line</i>
            </p>
          </>
        }
      />,
    );

    const inlineElement = screen.getByTestId('alertInline');
    expect(inlineElement).toBeInTheDocument();

    const title = screen.getByText('Default Banner with icon and title');
    expect(title).toBeInTheDocument();

    const subtextOne = screen.getByText((content, element) => {
      return element?.textContent === 'This is the first line';
    });
    expect(subtextOne).toBeInTheDocument();

    const subtextTwo = screen.getByText((content, element) => {
      return element?.textContent === 'This is the second line';
    });
    expect(subtextTwo).toBeInTheDocument();

    const strongText = screen.getByText('first line');
    expect(strongText.tagName).toBe('STRONG');

    const italicText = screen.getByText('second line');
    expect(italicText.tagName).toBe('I');

    const icon = screen.getByTestId('icon');
    expect(icon).toHaveAttribute('data-icon', 'circle');
  });

  test('renders full width alert when inline', () => {
    render(
      <Alert
        variant="Inline"
        state="Success"
        title="Full Width Alert!"
        subtext={<p>This alert takes the full width.</p>}
        isFullWidth
      />,
    );

    const bannerElement = screen.getByTestId('alertInline');
    expect(bannerElement).toBeInTheDocument();
    const alert = screen.getByText('Full Width Alert!');
    expect(alert).toHaveClass('w-full');
  });

  test('renders with custom background color', () => {
    render(
      <Alert
        variant="Banner"
        state="Success"
        title="Custom Background Color Alert!"
        subtext={<p>This alert has a custom background color.</p>}
        alertBackgroundColor="bg-colour-primary-light"
      />,
    );
    const alert = screen.getByTestId('alert');
    expect(alert).toHaveClass('bg-colour-primary-light');
  });

  test('renders with custom icon accent color', () => {
    render(
      <Alert
        variant="Banner"
        state="Success"
        title="Custom Icon Accent Color Alert!"
        subtext={<p>This alert has a custom icon accent color.</p>}
        iconAccentColor="text-colour-onSurface-light"
      />,
    );

    const icon = screen.getByTestId('alert-icon');
    expect(icon).toHaveClass('text-colour-onSurface-light');
  });

  test('renders children as a button and handles click events', () => {
    const handleClick = jest.fn();

    render(
      <Alert variant="Banner" title="This is a banner with a button" state="Success">
        <Button type="button" onClick={handleClick}>
          Click me
        </Button>
      </Alert>,
    );

    const button = screen.getByRole('button', { name: /Click me/i });
    expect(button).toBeInTheDocument();

    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
