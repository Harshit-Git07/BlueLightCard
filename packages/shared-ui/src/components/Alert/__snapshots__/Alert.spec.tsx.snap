// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Component renders correctly for Banner variant 1`] = `
<div
  className="
        false 
        w-full
        bg-colour-success-bright-light dark:bg-colour-success-container-dark 
        sticky top-0 left-0 z-50 px-[1rem] tablet:px-[3.5rem] laptop:px-[6.5rem] py-[1rem] 
        flex-col rounded 
        justify-between 
      "
  data-testid="alert"
>
  <div
    className="flex basis-full flex-row"
    data-testid="alertBanner"
  >
    <div
      className="flex items-start tablet:items-center"
    >
      <div
        className="text-colour-success-light dark:text-colour-success-dark"
        data-testid="alert-icon"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-circle-check text-[20px] flex content-start pt-1 tablet:pt-[0px] tablet:text-[24px]"
          data-icon="circle-check"
          data-prefix="fas"
          data-testid="icon"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM369 209L241 337c-9.4 9.4-24.6 9.4-33.9 0l-64-64c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l47 47L335 175c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </div>
    </div>
    <div
      className="tablet:flex tablet:mx-6 mx-4 m-auto tablet:justify-between w-full"
    >
      <div
        className="flex-col content-center tablet:pr-[16px] max-w-[750px]"
      >
        <p
          className="w-full text-colour-onSurface-light dark:text-colour-onSurface-dark break-words font-typography-title-small font-typography-title-small-weight text-typography-title-small leading-typography-title-small tracking-typography-title-small tablet:font-typography-title-medium-semibold tablet:font-typography-title-medium-semibold-weight tablet:text-typography-title-medium-semibold tablet:leading-typography-title-medium-semibold tablet:tracking-typography-title-medium-semibold"
        >
          This is a banner alert!
        </p>
        <div
          className="flex-col w-full pt-2 font-typography-body-small font-typography-body-small-weight text-typography-body-small leading-typography-body-small tracking-typography-body-small tablet:font-typography-body tablet:font-typography-body-weight tablet:text-typography-body tablet:leading-typography-body tablet:tracking-typography-body text-colour-onSurface-subtle-light dark:text-colour-onSurface-subtle-dark break-words"
        >
          <p>
            Operation completed successfully.
          </p>
        </div>
      </div>
    </div>
    <div
      className="flex items-start  tablet:items-center"
    >
      <button
        aria-label="Close"
        className="content-center text-center rounded-full text-colour-onSurface-subtle-light dark:text-colour-onSurface-subtle-dark"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-xmark text-[20px]"
          data-icon="xmark"
          data-prefix="far"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M345 137c9.4-9.4 9.4-24.6 0-33.9s-24.6-9.4-33.9 0l-119 119L73 103c-9.4-9.4-24.6-9.4-33.9 0s-9.4 24.6 0 33.9l119 119L39 375c-9.4 9.4-9.4 24.6 0 33.9s24.6 9.4 33.9 0l119-119L311 409c9.4 9.4 24.6 9.4 33.9 0s9.4-24.6 0-33.9l-119-119L345 137z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Alert Component renders correctly for Inline variant 1`] = `
<div
  className="
        false 
        mx-auto
        bg-colour-info-bright-light dark:bg-colour-info-bright-dark 
        relative  px-[0.75rem] py-[0.75rem] 
        flex-col rounded 
        justify-between 
      "
  data-testid="alert"
>
  <div
    className="flex basis-full flex-row"
    data-testid="alertInline"
  >
    <div
      className="flex items-start "
    >
      <div
        className="text-colour-onInfo-container-light dark:text-colour-onInfo-container-dark"
        data-testid="alert-icon"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-circle-info text-[20px] flex content-start pt-1 tablet:pt-[5px]"
          data-icon="circle-info"
          data-prefix="fas"
          data-testid="icon"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </div>
    </div>
    <div
      className=" mx-4 m-auto tablet:justify-between w-full"
    >
      <div
        className="flex-col content-center tablet:pr-[16px] max-w-[750px]"
      >
        <p
          className="w-full text-colour-onSurface-light dark:text-colour-onSurface-dark break-words font-typography-body-small-semibold font-typography-body-small-semibold-weight text-typography-body-small-semibold leading-typography-body-small-semibold tracking-typography-body-small-semibold  "
        >
          This is an inline alert!
        </p>
        <div
          className="flex-col w-full pt-1 font-typography-body-small font-typography-body-small-weight text-typography-body-small leading-typography-body-small tracking-typography-body-small text-colour-onSurface-subtle-light dark:text-colour-onSurface-subtle-dark break-words"
        >
          <p>
            Some important information here.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
